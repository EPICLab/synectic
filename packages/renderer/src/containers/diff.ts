import {diffLines} from 'diff';

/**
 * Diffs two blocks of text, comparing line by line, based on the algorithm proposed
 * in "An O(ND) Difference Algorithm and its Variations" (Myers, 1986). Results are formatted
 * according to the unified diff patch used in the Git protocol, where +/- indicate lines
 * added or removed in order to transform from original to updated.
 * @param original The base version of code or text to evaluate.
 * @param updated The updated version of code or text to compare against original.
 * @returns {string} A diff patch that conforms to the unified diff patch format.
 */
export const diff = (original: string, updated: string): string => {
  const diffLineChanges = diffLines(original, updated);
  let readableDiff = '';

  diffLineChanges.map(segment => {
    const segmentLines = segment.value.split(/\n/g, segment.count);
    if (segment.added) {
      segmentLines.map(line => (readableDiff += `+${line}\n`));
    } else if (segment.removed) {
      segmentLines.map(line => (readableDiff += `-${line}\n`));
    } else {
      segmentLines.map(line => (readableDiff += ` ${line}\n`));
    }
  });

  return readableDiff;
};
