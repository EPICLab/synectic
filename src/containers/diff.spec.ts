import { diff } from './diff';

describe('diff', () => {
  const good = 'Good dog\nGood dog\nBad dog';
  const bad = 'Bad dog\nGood dog\nGood dog';

  it('diff matches unmodified text without line changes', () => {
    expect(diff(good, good)).toMatchSnapshot();
  });

  it('diff returns both added and removed lines', () => {
    expect(diff(good, bad)).toMatchSnapshot();
  });
});