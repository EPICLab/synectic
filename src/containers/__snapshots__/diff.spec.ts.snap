// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diff diff matches unmodified text without line changes 1`] = `
" Good dog
 Good dog
 Bad dog
"
`;

exports[`diff diff returns both added and removed lines 1`] = `
"+Bad dog
 Good dog
-Good dog
-Bad dog
+Good dog
"
`;
